src/python/pants/backend/codegen/thrift/jvm_symbol_mapper.py

```diff
@@ -0,0 +1,66 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import DefaultDict, Mapping
+
+from pants.backend.codegen.thrift.target_types import AllThriftTargets, ThriftSourceField
+from pants.backend.codegen.thrift.thrift_parser import ParsedThrift, ParsedThriftRequest
+from pants.engine.addresses import Address
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.jvm.dependency_inference.artifact_mapper import MutableTrieNode
+from pants.jvm.dependency_inference.symbol_mapper import SymbolMap
+from pants.jvm.subsystems import JvmSubsystem
+from pants.jvm.target_types import JvmResolveField
+from pants.util.ordered_set import OrderedSet
+
+_ResolveName = str
+
+
+@dataclass(frozen=True)
+class FirstPartyThriftJvmMappingRequest:
+    extra_lang_ids: tuple[str, ...] = ()
+    extra_namespace_directives: tuple[str, ...] = ()
+
+
+@rule
+async def map_first_party_thirft_targets_to_jvm_symbols(
+    request: FirstPartyThriftJvmMappingRequest, thrift_targets: AllThriftTargets, jvm: JvmSubsystem
+) -> SymbolMap:
+    jvm_thrift_targets = [tgt for tgt in thrift_targets if tgt.has_field(JvmResolveField)]
+
+    parsed_thrift_sources = await MultiGet(
+        Get(
+            ParsedThrift,
+            ParsedThriftRequest(
+                sources_field=tgt[ThriftSourceField],
+                extra_namespace_directives=request.extra_namespace_directives,
+            ),
+        )
+        for tgt in jvm_thrift_targets
+    )
+
+    package_symbols: DefaultDict[tuple[_ResolveName, str], OrderedSet[Address]] = defaultdict(
+        OrderedSet
+    )
+    lang_ids = {"java", *request.extra_lang_ids}
```

Comment:
IIRC, this will only be safe if the java and scala code are generated into different namespaces. That is best practice afaik, so maybe this is a safe default. But I could imagine someone needing to configure this.

Since I'm low on context, it should be ok to leave this for now and wait until someone asks for a setting.

Reply:
yeah, if someone defines Java and Scala code in the same packages there is going to be a conflict. The reasoning for having this is mostly down to the fact that it should be possible to use Java generated classes in Scala.